### PR TITLE
Task 1

### PR DESCRIPTION
**Pull request by ChoCoBo1712**
Completed second module tasks

### CODE REVIEW

File: repository/src/main/java/com/epam/esm/repository/impl/ClauseBuilder.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  ClauseBuilder(String start, String separator) {
+    this.start = start;
+    this.separator = separator;
+  }
+
+  void addPart(String part) {
+    parts.add(part);
+  }
+
+  String build() {
```
Comment (eyaskevich26) [inline repository/src/main/java/com/epam/esm/repository/impl/ClauseBuilder.java:23]: I suppose the method can be shortened to:
`return String.join(separator, parts) + NEW_LINE;`

File: repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    @Override
+    public void addRelation(long certificateId, long tagId) {
+        SqlParameterSource parameters = new MapSqlParameterSource()
+                .addValue(CERTIFICATE_ID_PARAM, certificateId)
+                .addValue(TAG_ID_PARAM, tagId);
+        jdbcTemplate.update(INSERT_CERTIFICATES_TAGS_RELATION, parameters);
+    }
+
+    @Override
+    public void removeRelation(long certificateId, long tagId) {
```
Comment (ShiniTada) [inline repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: make the name of the method more specific. removeRelation(long, long) is super abstract. From the first glance it's difficult to realize what relation is this method about. 
what would you do if certificate has 2/3 relations?
the same in addRelation

File: service/src/main/java/com/epam/esm/dto/GiftCertificateDto.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+  @JsonFormat(
+      shape = JsonFormat.Shape.STRING,
+      pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
+      timezone = "UTC")
+  private Instant lastUpdateDate;
+
+  private Set<TagDto> tags;
+
+  public static GiftCertificateDto convertToDto(GiftCertificate giftCertificate, Set<Tag> tags) {
```
Comment (eyaskevich26) [inline service/src/main/java/com/epam/esm/dto/GiftCertificateDto.java]: It is better to separate conversion logic from the data object, e.g. to the `GiftCertificateConverter` with methods `convertToEntity(Dto dto)`, `convertToDto(...)`

File: service/src/main/java/com/epam/esm/dto/TagDto.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import org.modelmapper.ModelMapper;
+
+@Data
+public class TagDto {
+
+  private long id;
+
+  private String name;
+
+  public static TagDto convertToDto(Tag tag) {
```
Comment (eyaskevich26) [inline service/src/main/java/com/epam/esm/dto/TagDto.java]: The same as for `GiftCertificateDto`.

File: repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    @Override
+    public Optional<Tag> findByName(String name) {
+        SqlParameterSource parameters = new MapSqlParameterSource().addValue(NAME_PARAM, name);
+        List<Tag> tags = jdbcTemplate.query(SELECT_BY_NAME, parameters, rowMapper);
+        return Optional.ofNullable(tags.size() != 0 ? tags.get(0) : null);
+    }
+
+    @Override
+    public Set<Tag> findByCertificateId(long certificateId) {
```
Comment (ShiniTada) [inline repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java:94]: why TagRepositoryImpl has method Set<Tag> findByCertificateId(long certificateId) ?

File: service/src/main/java/com/epam/esm/exception/EntityNotFoundException.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.epam.esm.exception;
+
+public class EntityNotFoundException extends RuntimeException {
+
+  private final long entityId;
```
Comment (ShiniTada) [inline service/src/main/java/com/epam/esm/exception/EntityNotFoundException.java:5]: why do you use Class<?> causeEntity in InvalidEntityException, but here don't?

File: service/src/main/java/com/epam/esm/service/GiftCertificateService.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+  private final GiftCertificateValidator giftCertificateValidator;
+  private final TagValidator tagValidator;
+  private final SortStringValidator sortStringValidator;
+
+  public GiftCertificateService(
+      GiftCertificateRepository giftCertificateRepository,
+      TagRepository tagRepository,
+      GiftCertificateValidator giftCertificateValidator,
+      TagValidator tagValidator,
+      SortStringValidator sortStringValidator) {
```
Comment (ShiniTada) [inline service/src/main/java/com/epam/esm/service/GiftCertificateService.java]: consider using @RequiredArgsConstructor as long as you use lombok library

File: repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java (hunk 2 above)
Comment (ChoCoBo1712) [reply repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: fixed this issue

File: repository/src/main/java/com/epam/esm/repository/impl/ClauseBuilder.java (hunk 1 above)
Comment (ChoCoBo1712) [reply repository/src/main/java/com/epam/esm/repository/impl/ClauseBuilder.java:23]: fixed

File: service/src/main/java/com/epam/esm/dto/GiftCertificateDto.java (hunk 3 above)
Comment (ChoCoBo1712) [reply service/src/main/java/com/epam/esm/dto/GiftCertificateDto.java]: done

File: service/src/main/java/com/epam/esm/dto/TagDto.java (hunk 4 above)
Comment (ChoCoBo1712) [reply service/src/main/java/com/epam/esm/dto/TagDto.java]: done

File: service/src/main/java/com/epam/esm/exception/EntityNotFoundException.java (hunk 6 above)
Comment (ChoCoBo1712) [reply service/src/main/java/com/epam/esm/exception/EntityNotFoundException.java:5]: During gift certificate creation or update user can mess up either in certificate name or in tag name. In InvalidEntityException Class<?> causeEntity helps to specify where the mistake occurred. When user searches for entity EntityNotFoundException can only be thrown for an exact entity that user was searching for. So there is no need to specify class

File: service/src/main/java/com/epam/esm/service/GiftCertificateService.java (hunk 7 above)
Comment (ChoCoBo1712) [reply service/src/main/java/com/epam/esm/service/GiftCertificateService.java]: done

**Pull request merged**